modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ESubNullSafeCallInvoke.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless.node;
+
+import org.elasticsearch.painless.Globals;
+import org.elasticsearch.painless.Locals;
+import org.elasticsearch.painless.Location;
+import org.elasticsearch.painless.MethodWriter;
+import org.objectweb.asm.Label;
+
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Implements a call who's value is null if the prefix is null rather than throwing an NPE.
+ */
+public class ESubNullSafeCallInvoke extends AExpression {
+    /**
+     * The expression gaurded by the null check. Required at construction time and replaced at analysis time.
+     */
+    private AExpression guarded;
+
+    public ESubNullSafeCallInvoke(Location location, AExpression guarded) {
+        super(location);
+        this.guarded = requireNonNull(guarded);
+    }
+
+    @Override
+    void extractVariables(Set<String> variables) {
+        guarded.extractVariables(variables);
+    }
+
+    @Override
+    void analyze(Locals locals) {
+        guarded.expected = expected;
```

Comment:
I still need to think about this some more, but let me try to explain a bit better the implications of each variable being used here...

expected -- This can definitely be null. I tried to write an explanation in the package-info.  The gist of it is that it will be null in cases where promotion is required since we can't determine what expected is until after knowing both the right-hand and left-hand types (I think we want this to always be null here actually, explained later)

explicit -- This is set by an explicit cast and indicates that certain def cases cannot be optimized if it's true.  (Since expected will always be null this doesn't need to be set)

internal -- This should not be set here as it's really only used for arguments being passed into a function call where arguments may need to be boxed beyond the user's ability to do so.  Take a look at PSubCall or PSubDefCall for an example of this.  (Auto boxing will only work if expected and actual are set to the appropriate types which we cannot necessarily know.)

actual -- The type returned from the guarded node.  (Set correctly after analysis)

I don't think you ever want to cast here because of the way this node is a wrapper or middle man, so you are sort of double casting since the parent node would do the same cast in theory.  I hope that kind of makes sense.  For the case of boxing, I think this needs to be specialized for this node, so in analysis if you see the actual type is a primitive you want to make actual the boxed type, but track that you did that with a boolean in the node.  Then in writing make sure if you did box, you have to do it there too.